lib/compress/zstd_opt.c

```diff
@@ -423,237 +424,241 @@ size_t ZSTD_compressBlock_opt_generic(ZSTD_CCtx* ctx,
-                 * litLength == 0, where offset 0 means something special.
-                 * mlen == 1 means the previous byte was stored as a literal,
-                 * so they are mutually exclusive.
-                 */
-                assert(!(opt[cur].off == ZSTD_REP_MOVE_OPT && mlen == 1));
-                opt[cur].rep[0] = (opt[cur].off == ZSTD_REP_MOVE_OPT) ? (opt[cur-mlen].rep[0] - 1) : (opt[cur-mlen].rep[opt[cur].off]);
-           }
+            /* last match must start at a minimum distance of 8 from oend */
+            if (inr > ilimit) continue;
+
+            /* update repcodes */
+            {   U32 const mlen = opt[cur].mlen;
+                if (opt[cur].off > ZSTD_REP_MOVE_OPT) {
+                    opt[cur].rep[2] = opt[cur-mlen].rep[1];
+                    opt[cur].rep[1] = opt[cur-mlen].rep[0];
+                    opt[cur].rep[0] = opt[cur].off - ZSTD_REP_MOVE_OPT;
+                } else {
+                    opt[cur].rep[2] = (opt[cur].off > 1) ? opt[cur-mlen].rep[1] : opt[cur-mlen].rep[2];
+                    opt[cur].rep[1] = (opt[cur].off > 0) ? opt[cur-mlen].rep[0] : opt[cur-mlen].rep[1];
+                    /* If opt[cur].off == ZSTD_REP_MOVE_OPT, then mlen != 1.
+                     * offset ZSTD_REP_MOVE_OPT is used for the special case
+                     * litLength == 0, where offset 0 means something special.
+                     * mlen == 1 means the previous byte was stored as a literal,
+                     * so they are mutually exclusive.
+                     */
+                    assert(!(opt[cur].off == ZSTD_REP_MOVE_OPT && mlen == 1));
+                    opt[cur].rep[0] = (opt[cur].off == ZSTD_REP_MOVE_OPT) ? (opt[cur-mlen].rep[0] - 1) : (opt[cur-mlen].rep[opt[cur].off]);
+            }   }
 
             best_mlen = minMatch;
-            {   U32 i, last_i = ZSTD_REP_CHECK + (mlen != 1);
-                for (i=(opt[cur].mlen != 1); i<last_i; i++) {  /* check rep */
-                    const S32 repCur = (i==ZSTD_REP_MOVE_OPT) ? (opt[cur].rep[0] - 1) : opt[cur].rep[i];
-                    if ( (repCur > 0) && (repCur < (S32)(inr-prefixStart))
-                       && (ZSTD_readMINMATCH(inr, minMatch) == ZSTD_readMINMATCH(inr - repCur, minMatch))) {
-                       mlen = (U32)ZSTD_count(inr+minMatch, inr+minMatch - repCur, iend) + minMatch;
-
-                       if (mlen > sufficient_len || cur + mlen >= ZSTD_OPT_NUM) {
-                            best_mlen = mlen; best_off = i; last_pos = cur + 1;
-                            goto _storeSequence;
-                       }
+            {   U32 const ll0 = (opt[cur].mlen != 1);
+                U32 const lastR = ZSTD_REP_CHECK + ll0;
+                U32 repCode4;  /* universal referential */
+                for (repCode4=ll0; repCode4<lastR; repCode4++) {  /* check rep */
+                    const S32 repCur = (repCode4==ZSTD_REP_MOVE_OPT) ? (opt[cur].rep[0] - 1) : opt[cur].rep[repCode4];
```

Comment:
You renamed this to `repOffset` above, do you want to do the same here?

Reply:
That's a relevant point.
Thing is, this code will disappear in next update.